### DISCUSSION TITLE
Button margin control

### DISCUSSION THREAD

**Discussion by franzvezuli** [beercss/beercss#329] [Q&A] [answered]
Hello!

First of all, I want to say how much I appreciate this project! Thank you so much for your efforts.

I have a question, why does the `Button` element have a default padding on the x-axis? I'm building a simple login card, example here: https://codepen.io/franzvz/pen/ZEdELRG

As you can see, so far I just have
```
<button class="small-round">Button</button>
```

But the form doesn't look too nice, I wanted to give it a but of a breathing room, so I wanted to add some padding like so: https://codepen.io/franzvz/pen/XWLWpBK

```
<button class="small-round medium-margin top-margin">Sign in</button>
```

Looks a bit better, but I noticed the button for some reason was not left aligned.

Not a deal breaker, but I have two questions:
1. Why do button elements have any opinion on default x-axis padding? Is this a Material Design spec thing?
2. Button padding recommendations?

It seems you can't really mix-match the margin helpers to create something like this:
```
no-padding left-padding medium-margin top-margin
```
which makes sense, there should be no such thing as class ordering logic. My work around to left align the whole login form is just to add `style="margin-left: 0"` to the button element so far.

To be honest I don't even know which one looks better, but its just something I noticed.

Thank you for your time!

**Comment by leonardorafael** [comment]
Hi @franzvezuli ,

In semantic HTML is a good practice to have a `button` inside a `nav`. Because `nav` means a collection of clickable elements. Try this:

```html
<nav>
  <button class="small-round">Sign in</button>
</nav>
```

Using the semantic HTML in a right way, will avoid to use helpers like `*-margin` and `*-padding` in all things. And you will get a clean code as result. Additionally, you can use the `*-space` and `*-align` helpers on `nav`:
https://github.com/beercss/beercss/blob/main/docs/HELPERS.md#spaces
https://github.com/beercss/beercss/blob/main/docs/HELPERS.md#alignments

**Comment by leonardorafael** [comment]
> Not a deal breaker, but I have two questions:
> 
> 1. Why do button elements have any opinion on default x-axis padding? Is this a Material Design spec thing?
> 2. Button padding recommendations?

1. It's a MD3 spec https://m3.material.io/components/buttons/specs
2. The default button has the padding like MD3 guidelines.

**Comment by fvezulimo** [comment]
@leonardorafael Thank you so much for the quick reply. I tried that and it looks good! A few follow ups though that I had:

> Because `nav` means a collection of clickable elements.

That's interesting, I've never heard it describe like that in the context of non-navigational links.
https://www.w3schools.com/tags/tag_nav.asp describes it as `The <nav> tag defines a set of navigation links.`
https://learntheweb.courses/topics/html-semantics-cheat-sheet/ has it as `<nav> Defines a group a navigation links.`

The buttons would usually go in a `<form>` element are usually for actions (optionally driven by JS).

> 1. It's a MD3 spec https://m3.material.io/components/buttons/specs
> 2. The default button has the padding like MD3 guidelines.

I'm reading the spec now, are you referring to this part?

> Padding between elements 8dp

<details>
<summary>Screenshot</summary>
<img width="1241" alt="image" src="https://github.com/beercss/beercss/assets/127247895/a5f57872-6b8f-4df8-bca3-ae161cb9ffa0">
</details>

That seems to be for elements that are inside the button itself, but not elements outside of the button.

**Comment by leonardorafael** [comment]
Oh do you mean margins instead paddings.

We decided to add margins on buttons because you can add two or more buttons (side by side) without need to set a margin in all of them (can this be the most use case? maybe).

We recommend to use `nav` or `row` as container for the clickable elements, this will help to add some helpers and for a cleaner code.

Some people doesn't like to use `nav` with buttons (there are many different opinions about that), so you can use `row` instead too.

Hope this helps.

**Comment by leonardorafael** [comment]
> ![image](https://github.com/beercss/beercss/assets/12979016/ba769308-641e-48f3-b0b9-4cf1115f01e0)

This MD3 image is curious, "Padding between elements" is new to me 😅

**Comment by fvezulimo** [comment]
Thanks for the answers! Yes I meant margins, not padding, sorry about that.

The image was me trying to find where you saw this

> The default button has the padding like MD3 guidelines.
And this is the only mention of `padding` in the doc

So for margins

> We decided to add margins on buttons because you can add two or more buttons (side by side) without need to set a margin in all of them (can this be the most use case? maybe).

This was just a design decision for Beer CSS, not necessarily because of the MD3 spec then. Which is totally fine! thanks for the clarification.

---

> Some people doesn't like to use nav with buttons (there are many different opinions about that), so you can use row instead too.

I don't like using `nav` for `Form Buttons`. `row` seems interesting, but I don't think it is an HTML element:
```
<row>
  <button class="small-round">Sign in</button>
</row>
```
does not seem to work, but
```
<div class="row">
  <button class="small-round">Sign in</button>
</div>
```
or even
```
<button class="row small-round">Sign in</button>
```
by itself seem to work just fine.

Is `row` a Beer CSS helper or an element? If its a helper, maybe we have to update https://github.com/beercss/beercss/blob/main/docs/HELPERS.md because I don't see it mention there, but in my codepen `class="row"` seems to also fix the problem.

**Comment by leonardorafael** [comment, marked as answer]
> Is row a Beer CSS helper or an element?

Hi, `row` is an element. `row` is basically an alias for `nav` element. The elements in Beer CS can be a tag or a class.

The documentation is here (maybe we can adjust here to be more intuitive):
https://github.com/beercss/beercss/blob/main/docs/NAVIGATION.md

**Comment by franzvezuli** [comment]
> The elements in Beer CS can be a tag or a class

That's not what I'm seeing though 
```
<!-- Does not work -->
<row>
  <button class="small-round">Sign in</button>
</row>

<!-- Works -->
<div class="row">
  <button class="small-round">Sign in</button>
</div>
```

![image](https://github.com/beercss/beercss/assets/4238855/8ba62454-68be-4de3-b977-1f9fecb2c0e8)

Only the second one removes the margins: https://codepen.io/franzvz/pen/MWMWGZv

Class works, but not the tag.
